lib/matplotlib/axes/_axes.py

```diff
@@ -753,6 +753,164 @@ def text(self, x, y, s, fontdict=None, **kwargs):
+        # Get last data point for each handle
+        xys = list(map(get_last_data, handles))
+
+        # Find the largest last x value
+        data_maxx = max(map(lambda point: point[0], xys))
+        # TODO(omarchehab98): We do eventually sort `xys` by `y`.
+        #                     If we reorder things we can do these two
+        #                     computations in constant time.
+        #                     Rather than linear time.
+        # Find the smallest last y value
+        data_miny = min(map(lambda point: point[1], xys))
+        # Find the largest last y value
+        data_maxy = max(map(lambda point: point[1], xys))
+
+        # Get the figure width and height in pixels
+        fig_dpi_transform = self.figure.dpi_scale_trans.inverted()
+        fig_bbox = self.get_window_extent().transformed(fig_dpi_transform)
+        fig_width_px = fig_bbox.width * self.figure.dpi
+        fig_height_px = fig_bbox.height * self.figure.dpi
+
+        # Get the figure width and height in points
+        fig_minx, fig_maxx = self.get_xbound()
+        fig_miny, fig_maxy = self.get_ybound()
+        fig_width_pt = abs(fig_maxx - fig_minx)
+        fig_height_pt = abs(fig_maxy - fig_miny)
+
+        # Space to the left of the text converted from pixels
+        margin_left = 8 * (fig_width_pt / fig_width_px)
+        # Space to the top and bottom of the text converted from pixels
+        margin_vertical = 2 * (fig_height_pt / fig_height_px)
+
+        text_fontsize = 10
+        # Height of the text converted from pixels
+        text_height = text_fontsize * (fig_height_pt / fig_height_px)
+
+        # Height of each bucket
+        bucket_height = text_height + 2 * margin_vertical
+        # Total number of buckets
+        buckets_total = 1 + int((fig_maxy - fig_miny - text_height) /
+                                 bucket_height)
+        # Bit array to track which buckets are used
+        buckets_map = 0
+
+        def get_bucket_index(y):
+            return int((y - fig_miny) / bucket_height)
+
+        # How many text SHOULD be in this bucket
+        bucket_densities = [0] * buckets_total
+        for xy in xys:
+            data_x, data_y = xy
+            bucket = get_bucket_index(data_y)
+            if bucket >= 0 and bucket < buckets_total:
+                bucket_densities[bucket] += 1
+
+        def in_viewport(args):
+            xy = args[2]
+            x, y = xy
+            return fig_minx < x < fig_maxx and fig_miny < y < fig_maxy
+
+        def by_y(args):
+            xy = args[2]
+            x, y = xy
+            return y
+
+        bucket_offset = None
+        prev_ideal_bucket = -1
+        # Iterate over all the handles, labels, and xy data sorted by y asc
+        for handle, label, xy in sorted(filter(in_viewport,
+                                               zip(handles, labels, xys)),
+                                        key=by_y):
+            data_x, data_y = xy
+
```

Comment:
Can you explain this?  Why is this more complicated than putting the text at the end of each line?

Reply:
Our main goal with this implementation is to ensure readability for the user at minimal cost for them. The given examples such as Bachelor's degrees by gender were created by individually placing text objects and using offsets so that they do not overlap. Therefore to try and automatically format the texts in such a way that there is no overlap, we decided to create this bitmap system for spacing rather than simply placing a text at the end of each line. This need for automatic spacing is exemplified by the the same graph from the feature request, where it lacks the offsets.
![image](https://user-images.githubusercontent.com/25535247/78528562-aafd5c80-77ad-11ea-8b2b-cf7cb2e24d2e.png)

Reply:
@jklymak basically, if you place the text at the end of each line. In some cases the text will overlap, consider the bachelor's degree example posted by my colleague above.

Reply:
@jklymak for the explanation you requested:

We are placing annotations in "buckets" for text to be placed in.

Reply:
@jklymak for the explanation you requested:

We are placing annotations in "buckets". To do so, we sort the last data points by Y ascending and use a boolean array to keep track of which buckets have been used. If a bucket has been used, we try to place the text in the next available bucket.

We also consider the data point density of each bucket, if there are many days points in a bucket the annotations should be centered around the bucket rather than be placed bottom-up starting from the center (due to our iteration order because of the sort)